### PR TITLE
Exchange custom run_tests scripts with mozharness (#642)

### PR DESCRIPTION
**Pull request by whimboo**
This PR will fix issue #642. As of now it is not ready for a full review yet.  But maybe you already want to have a peak on it to give feedback? CC'ing @mjzffr and @sydvicious.

### CODE REVIEW

**Comment by whimboo** [comment]
I think you both can start reviewing this PR! The only thing which is left right now is an update for the update tests in regards the HTTP.log. Bug 1210753 I will fix in a follow-up issue.

File: jenkins-master/jobs/mozilla-aurora_functional/config.xml [hunk 1]
```diff
@@ -48,32 +48,9 @@
...
-        <localBranch>mozilla-aurora</localBranch>
-      </hudson.plugins.git.extensions.impl.LocalBranch>
-      <hudson.plugins.git.extensions.impl.CloneOption>
-        <shallow>true</shallow>
-        <reference></reference>
-      </hudson.plugins.git.extensions.impl.CloneOption>
-      <hudson.plugins.git.extensions.impl.RelativeTargetDirectory>
-        <relativeTargetDir>firefox-ui-tests</relativeTargetDir>
-      </hudson.plugins.git.extensions.impl.RelativeTargetDirectory>
-    </extensions>
```
Comment (sydvicious) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: Right now, firefox-media-tests has firefox-ui-tests as a submodule. Will the mozharness action which calls get deal with submodules? Or is there some other mechanism for insuring that firefox-media-tests can find firefox-ui-tests?

File: jenkins-master/jobs/mozilla-aurora_functional/config.xml [hunk 2]
```diff
@@ -86,20 +63,22 @@
...
     <hudson.plugins.xshell.XShellBuilder plugin="xshell@0.9-SNAPSHOT">
-      <commandLine>python runtests.py --type=functional --installer-url=$INSTALLER_URL</commandLine>
+      <commandLine>python -u runtests.py --test-type=functional --repository=mozilla-aurora --installer-url=$INSTALLER_URL</commandLine>
       <executeFromWorkingDir>false</executeFromWorkingDir>
     </hudson.plugins.xshell.XShellBuilder>
   </builders>
   <publishers>
     <hudson.tasks.ArtifactArchiver>
-      <artifacts>build/upload/*.*</artifacts>
+      <artifacts>build/upload/**</artifacts>
```
Comment (sydvicious) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: Well, this was just a bug anyway. Good change.

File: jenkins-master/jobs/mozilla-aurora_functional/config.xml (hunk 1 above)
Comment (whimboo) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: The checkout is part of your mozharness script but not mozmill-ci. There is mozharness/external_tools/gittool.py for your git related usage.

File: jenkins-master/jobs/mozilla-aurora_functional/config.xml [hunk 3]
```diff
@@ -195,7 +174,6 @@ ${FAILED_TESTS}
       <timeoutType>absolute</timeoutType>
       <timeoutMinutesElasticDefault>3</timeoutMinutesElasticDefault>
     </hudson.plugins.build__timeout.BuildTimeoutWrapper>
-    <hudson.plugins.timestamper.TimestamperBuildWrapper plugin="timestamper@1.5.14"/>
```
Comment (sydvicious) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: Are we no longer outputting timestamps in the build log because our logs will have them when they are uploaded?

Comment (whimboo) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: The mozharness logger already outputs timestamps. So I don't see the need to have this information twice and stealing visible space per line.

File: jenkins-master/jobs/scripts/workspace/config.py [hunk 4]
```diff
@@ -31,10 +35,9 @@
                 'tier': 3,
             },
             'logs': {
-                # Currently we don't have a gecko.log because we log to the console (bug 1174766)
-                # 'gecko': os.path.join(here, 'build', 'upload', 'logs', 'gecko.log'),
-                'http.log': os.path.join(here, 'upload', 'logs', 'http.log'),
-                'tbpl.log': os.path.join(here, 'upload', 'logs', 'tbpl.log'),
+                'buildbot_text': os.path.join(here, 'build', 'upload', 'logs', 'log_info.log'),
+                'http.log': os.path.join(here, 'build', 'upload', 'logs', 'http.log'),
+                'report.html': os.path.join(here, 'build', 'upload', 'reports', 'report.html'),
```
Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/config.py]: So, if I add a media type, and I have more files than these, I would list them similarly here?

File: jenkins-master/jobs/scripts/workspace/runtests.py [hunk 5]
```diff
@@ -5,85 +5,84 @@
...
-            '--workspace', os.getcwd(),
-            '--log-tbpl', settings['logs']['tbpl.log'],
-        ]
+        command = [os.path.join('mozharness', 'scripts', settings['harness_script']),
+                   '--cfg', settings['harness_config'],
+                   '--firefox-ui-branch', args.repository,
+                   '--installer-url', args.installer_url,
+                   ]
 
-        if args.type == 'update':
```
Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: I don't like this at all. I would much rather have a bass class that exposes methods to get options, and you would subclass it for update, functional or other types as they come along (like media, for instance). This certainly works, but does not have that OO spirit.

File: jenkins-master/jobs/scripts/workspace/config.py (hunk 4 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/config.py]: Right. We could also change the property name and something more general like `artifacts`.

File: jenkins-master/jobs/scripts/workspace/runtests.py [hunk 6]
```diff
@@ -5,85 +5,84 @@
...
-
-            manifests = [firefox_puppeteer.manifest, firefox_ui_tests.manifest_functional]
-            command.extend(manifests)
-
-        retval = 0
+        env = copy.copy(os.environ)
+        for var in ENV_VARS_TO_PURGE:
+            env.pop(var, None)
 
         print('Calling command to execute tests: {}'.format(command))
```
Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: print? Shouldn't this be some kind of logger?

File: jenkins-master/jobs/scripts/workspace/runtests.py [hunk 7]
```diff
@@ -5,85 +5,84 @@
...
-        retval = subprocess.call(command)
+        subprocess.check_call(command, env=env)
 
+    except subprocess.CalledProcessError as e:
+        retval = e.returncode
```
Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: so, where is retval used? You immediately raise, but the finally clause does nothing with this variable. Seems like dead code. Do you want to do something with retval here?

Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: And if retval is not used, you don't need the except clause; you just need try-finally.

**Comment by sydvicious** [comment]
I like where this is going a lot. A few questions here and there.

File: jenkins-master/jobs/scripts/workspace/runtests.py (hunk 5 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: Ok, let me try to make that a bit better. It's also not my favorite actually.

File: jenkins-master/jobs/scripts/workspace/runtests.py [hunk 8]
```diff
@@ -5,85 +5,84 @@
...
 
-        command = [
-            'firefox-ui-update' if args.type == 'update' else 'firefox-ui-tests',
-            '--installer', installer_path,
-            '--workspace', os.getcwd(),
-            '--log-tbpl', settings['logs']['tbpl.log'],
-        ]
+        command = [os.path.join('mozharness', 'scripts', settings['harness_script']),
+                   '--cfg', settings['harness_config'],
+                   '--firefox-ui-branch', args.repository,
```
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: As discussed on IRC media tests do not have different branches at the moment. So this should definitely go into a customized test type section.

File: jenkins-master/jobs/mozilla-aurora_functional/config.xml (hunk 1 above)
Comment (mjzffr) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: Yes, the mozharness action deals with submodules. You need to configure the
revision of firefox-medie-tests and the revision of the firefox-ui-tests
submodule. Here's the code:
https://dxr.mozilla.org/mozilla-central/source/testing/mozharness/mozharness/mozilla/testing/firefox_media_tests.py#201

In any case, the submodule will go away once firefox_puppeteer is a
separate package in-tree.

Maja Frydrychowicz
irc: maja_zf

File: jenkins-master/jobs/scripts/workspace/runtests.py (hunk 5 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: I pushed a new commit which refactors runtests.py and introduces runner classes.

File: jenkins-master/jobs/scripts/workspace/runtests.py (hunk 6 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: >>          print('Calling command to execute tests: {}'.format(command))
> 
> print? Shouldn't this be some kind of logger?

We can do but it won't end-up in a log file we upload. If you think it
is important now I can do it. Otherwise we can file a follow-up issue
maybe mentored. Please let me know.

File: jenkins-master/jobs/scripts/workspace/runtests.py (hunk 7 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: >> +    except subprocess.CalledProcessError as e:
>> +        retval = e.returncode
> 
> so, where is retval used? You immediately raise, but the finally clause
> does nothing with this variable. Seems like dead code. Do you want to do
> something with retval here?

Well, in finally we write this exit code to retval.txt, which is then
picked up by submission.py in the post_build step for evaluation.

File: jenkins-master/jobs/mozilla-aurora_functional/config.xml (hunk 1 above)
Comment (whimboo) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: Why is the revision of the `firefox-ui-tests` repository not taken from the `firefox-media-tests` repository? Usually you use `git submodule init/update` but maybe gittool doesn't support it?

Comment (mjzffr) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: If I recall correctly, gittool doesn't support it.

Maja Frydrychowicz
irc: maja_zf

Comment (whimboo) [inline jenkins-master/jobs/mozilla-aurora_functional/config.xml]: I looked this up and it is true. So media jobs would simply have to specify another argument for runtests.py to accommodate that.

**Comment by whimboo** [comment]
The last commit covers the missing archive task and upload step to Treeherder for the http.log file.

**Comment by whimboo** [comment]
One thing I will do now is to add some more documentation as @mjzffr asked for. But that should not stop us from reviewing the functional pieces of this patch.

File: jenkins-master/jobs/scripts/workspace/config.py [hunk 9]
```diff
@@ -18,11 +20,13 @@
                 'tier': 3,
             },
             'logs': {
```
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/config.py]: Something for discussion here... Given that we do not only upload logs, maybe we want to name that `artifacts`?

Comment (mjzffr) [inline jenkins-master/jobs/scripts/workspace/config.py]: Would it help to distinguish items that are used for "text_log_summary" artifact versus "Job Info" artifact on Treeherder? (For reference: https://treeherder.mozilla.org/api/project/mozilla-central/artifact/?job_id=2716806)

Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/config.py]: We could have a `raw_log` property for the `buildbot_text` to parse log, and `artifacts` for the rest?

Comment (mjzffr) [inline jenkins-master/jobs/scripts/workspace/config.py]: That sounds good.

**Comment by mjzffr** [comment]
I like the refactoring of runtests.py. Everything else looks good. Thanks for adding in all those doc strings, too. r+

File: jenkins-master/jobs/scripts/workspace/config.py (hunk 9 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/config.py]: I pushed a new commit with those changes. Means we have an `artifact` dict under the treeherder section now, which holds any kind of artifacts to be uploaded and to attach as job info. Beside that there is also a `log_reference` property which references the artifact which is used by TH for parsing the results.

Here the results from a test run:
https://treeherder.allizom.org/logviewer.html#?job_id=2340623&repo=mozilla-central

File: jenkins-master/jobs/mozilla-central_update/config.xml [hunk 10]
```diff
@@ -103,20 +80,22 @@ NSPR_LOG_FILE=http.log</propertiesContent>
...
     <hudson.plugins.xshell.XShellBuilder plugin="xshell@0.9-SNAPSHOT">
-      <commandLine>python runtests.py --type=update --installer-url=$INSTALLER_URL --update-channel=$CHANNEL --update-target-build-id=$TARGET_BUILD_ID</commandLine>
+      <commandLine>python -u runtests.py --test-type=update --repository=mozilla-central --installer-url=$INSTALLER_URL --update-channel=$CHANNEL --update-target-build-id=$TARGET_BUILD_ID</commandLine>
       <executeFromWorkingDir>false</executeFromWorkingDir>
     </hudson.plugins.xshell.XShellBuilder>
   </builders>
   <publishers>
     <hudson.tasks.ArtifactArchiver>
-      <artifacts>build/upload/*.*</artifacts>
+      <artifacts>build/http.log,build/upload/**</artifacts>
```
Comment (sydvicious) [inline jenkins-master/jobs/mozilla-central_update/config.xml]: So, why is this change necessary? Did something start writing logs into a different place? Or is this just narrowing what we archive? (Jenkins archiving gets really poor performance with really large directories).

File: jenkins-master/jobs/scripts/workspace/config.py [hunk 11]
```diff
@@ -10,32 +10,38 @@
...
         'functional': {
+            'harness_config': os.path.join('firefox_ui_tests', 'qa_jenkins.py'),
+            'harness_script': os.path.join('firefox_ui_tests', 'functional.py'),
             'treeherder': {
                 'group_name': 'Firefox UI Tests - functional',
                 'group_symbol': 'Ff',
                 'job_name': 'Firefox UI Tests - functional ({locale})',
                 'job_symbol': '{locale}',
                 'tier': 3,
+                'artifacts': {
```
Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/config.py]: Much better name.

File: jenkins-master/jobs/scripts/workspace/runtests.py [hunk 12]
```diff
@@ -5,108 +5,141 @@
...
-here = os.path.dirname(os.path.abspath(__file__))
-venv_path = os.path.join(here, 'tests_venv')
+from config import config
+from jenkins import JenkinsDefaultValueAction
 
+# Purge unwanted environment variables like credentials
+ENV_VARS_TO_PURGE = [
+    'AWS_BUCKET', 'AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY',
+    'TREEHERDER_URL', 'TREEHERDER_CLIENT_ID', 'TREEHERDER_SECRET',
+]
```
Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: This is not flake8 or pep style. Do we care?

File: jenkins-master/jobs/mozilla-central_update/config.xml (hunk 10 above)
Comment (whimboo) [inline jenkins-master/jobs/mozilla-central_update/config.xml]: The http.log generation is setup in the jenkins job. See https://github.com/mozilla/mozmill-ci/pull/698/files#diff-fb5718bbd1cd8a1f8df92e49038df6c2L58. I can try to modify the location so it also ends-up in the upload folder. Not sure when this is created by mozharness (most likely at the end when logs are getting copied into that folder) and if Firefox is smart enough to create the full path. I will check.

Comment (sydvicious) [inline jenkins-master/jobs/mozilla-central_update/config.xml]: Why is this only in mozilla-central_update and not mozilla-aurora_update? Or is it and I am misreading?

Comment (whimboo) [inline jenkins-master/jobs/mozilla-central_update/config.xml]: This is only important for update tests and investigation of download problems of the MAR files.

So it will not work with a simple change in the job config. I would have to specify the path to the log file in a platform independent way which is not possible here. Option is to move the setting of the environment variables into the runtests.py script and have an extra section in config.py for additional env variables.

**Comment by sydvicious** [comment]
I am confident that everything is in order then. This is outstanding. An enthusiastic r+.

File: jenkins-master/jobs/scripts/workspace/runtests.py (hunk 12 above)
Comment (whimboo) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: Why should it not be? I don't see even a warning in my editor which automatically checks that when saving the file.

Comment (sydvicious) [inline jenkins-master/jobs/scripts/workspace/runtests.py]: I was under the impression that there was supposed to be one value per line. Not a big deal, so let's not stress about it.

File: jenkins-master/jobs/mozilla-central_update/config.xml (hunk 10 above)
Comment (whimboo) [inline jenkins-master/jobs/mozilla-central_update/config.xml]: So I cannot move the `http.log` to `build/upload/` because a possible clobber step in mozharness could purge this file then. We don't have control about that from the CI level.

**Comment by whimboo** [comment]
Syd found one more issue when running the jobs on Windows. So I pushed a new commit with the fix.

**Comment by whimboo** [comment]
We are currently blocked on [bug 1226769](https://bugzilla.mozilla.org/show_bug.cgi?id=1226769) and it dependencies. mozharness does not properly work at the moment on Windows for custom py scripts.

**Comment by whimboo** [comment]
I pushed another commit which enables backups of possible minidump files in case of crashes. We don't want to upload them, but I think it would be good to archive them together with the build for later more detailed inspections.

Otherwise I'm waiting for the next merge from mozilla-inbound to mozilla-central which contains a couple of mozharness fixes to get our jobs working on Windows. I hope it will happen soon.

**Comment by whimboo** [comment]
Oh and here an [example log](https://treeherder.allizom.org/logviewer.html#?job_id=2358907&repo=mozilla-central) on Treeherder for crash analysis.

**Comment by whimboo** [comment]
While running some more tests I noticed that runtests.py does not always have a busted retval. Especially not if the script returns immediately due to missing arguments. So I think it is better to have the retval.txt logic directly under `main()`, so we cover every problem and do not report an unknown result which currently is even not rejected by Treeherder (see bug 1224360).

**Comment by whimboo** [comment]
@mjzffr and @sydvicious I pushed a bump of mozharness for all my latest fixes. With that upgrade our jobs should run on Windows without failures now. So please review all changes from today. This PR is complete now.

**Comment by sydvicious** [comment]
r+. Looks awesome.

**Comment by mjzffr** [comment]
r+ from me, too.
One (optional) suggestion.

**Comment by whimboo** [comment]
I pushed one more commit for the suggestion from @mjzffr. I got it nicely working with a self defined enum class. @mjzffr if that looks fine to you please tell me and I can get the PR merged to master.

**Comment by mjzffr** [comment]
r+!
